### PR TITLE
upgraded CUDA.jl to version 3.3.0

### PR DESCRIPTION
**Pull request by tomchor**
Closes https://github.com/CliMA/Oceananigans.jl/issues/1684

It apparently solved the issue for me.

### CODE REVIEW

**Comment by glwagner** [comment]
> Closes #1684
> 
> It apparently solved the issue for me.

Makes sense. Good there's a tagged release we can use.

It looks like the toolkit version needs to be updated on sverdrup. I'll look into that and hopefully tests will pass.

**Comment by tomchor** [comment]
It seems like tests will pass this time. 

FYI, I was forced to remove the instances of `@disallowscalar` because that's not included in `CUDA.jl` anymore, but I'm guessing this is okay since "scalar iteration is now disallowed by default" according to @maleadt.

I'm guessing we wanna bump to 0.58.3 after this since this fixed an important bug. Can any one confirm?

**Comment by glwagner** [comment]
Hmm. I believe that scalar operations are (or were) _specifically_ allowed globally in our `runtests.jl`, overriding any default. Thus `@disallowscalar` was used to specifically disallow scalar operations for specific statements covered by that macro, without disallowing them globally for all subsequent tests. I don't think it matters whether or not scalar iteration is allowed by default since we are manually controlling that setting.

Furthermore, I don't think removing `@disallowscalar` would cause tests to fail; rather, it means that those tests _will_ pass even if they have scalar operations (where previously they would not have passed).

Instead of using `@disallowscalar` we can flip the global state; eg we can replace

```julia
@disallowscalar statement
```

with

```julia
CUDA.allowscalar(false)
statement
CUDA.allowscalar(true)
```

Perhaps this is what we want?

**Comment by glwagner** [comment]
> I'm guessing we wanna bump to 0.58.3 after this since this fixed an important bug. Can any one confirm?

For sure bump!

File: test/test_grids.jl [hunk 1]
```diff
@@ -552,7 +552,7 @@ end
         grid = RegularRectilinearGrid(topology=topo, size=(3, 7, 9), x=(0, 1), y=(-π, π), z=(0, 2π))
 
         @test try
-            CUDA.@disallowscalar show(grid); println()
+            show(grid); println()
```
Comment (glwagner) [inline test/test_grids.jl:555]:
```suggestion
            CUDA.allowscalar(false)           
            show(grid); println()
            CUDA.allowscalar(true)
```

File: test/test_grids.jl [hunk 2]
```diff
@@ -593,7 +593,7 @@ end
             grid = VerticallyStretchedRectilinearGrid(architecture=arch, size=(1, 1, Nz-1), x=(0, 1), y=(0, 1), z_faces=collect(0:Nz).^2)
             
             @test try
-                CUDA.@disallowscalar show(grid); println()
+                show(grid); println()
```
Comment (glwagner) [inline test/test_grids.jl:596]:
```suggestion
            CUDA.allowscalar(false)           
            show(grid); println()
            CUDA.allowscalar(true)
```

File: test/test_grids.jl [hunk 3]
```diff
@@ -617,7 +617,7 @@ end
         grid = RegularLatitudeLongitudeGrid(size=(36, 32, 1), longitude=(-180, 180), latitude=(-80, 80), z=(0, 1))
     
         @test try
-            CUDA.@disallowscalar show(grid); println()
+            show(grid); println()
```
Comment (glwagner) [inline test/test_grids.jl:620]:
```suggestion
            CUDA.allowscalar(false)           
            show(grid); println()
            CUDA.allowscalar(true)
```

File: test/test_grids.jl [hunk 4]
```diff
@@ -639,7 +639,7 @@ end
         grid = ConformalCubedSphereFaceGrid(size=(10, 10, 1), z=(0, 1))
     
         @test try
-            CUDA.@disallowscalar show(grid); println()
+            show(grid); println()
```
Comment (glwagner) [inline test/test_grids.jl:642]:
```suggestion
            CUDA.allowscalar(false)           
            show(grid); println()
            CUDA.allowscalar(true)
```

**Comment by glwagner** [comment]
Some sandbox stuff might've been added by accident

File: sandbox/ocean_wind_mixing_and_convection.jl [hunk 5]
```diff
@@ -0,0 +1,102 @@
+using Random
```
Comment (glwagner) [inline sandbox/ocean_wind_mixing_and_convection.jl]: Don't think this file was intended to be part of this PR?

File: sandbox/suman_example.jl [hunk 6]
```diff
@@ -0,0 +1,69 @@
+using Random
```
Comment (glwagner) [inline sandbox/suman_example.jl:1]: Ditto for this file.

**Review by glwagner — APPROVED** [review]

**Comment by tomchor** [comment]
> Some sandbox stuff might've been added by accident

Ah, thanks for catching that. I was investigating the stretched grid + timestep wizard thing. Should be good now.

**Comment by maleadt** [comment]
> Hmm. I believe that scalar operations are (or were) _specifically_ allowed globally in our `runtests.jl`, overriding any default.

I was considering deprecating that option though, so that scalar iteration would be disabled by default and can only be allowed for a limited number of statements using `@allowscalar` or `allowscalar() do ... end`, because it's such a performance trap. Is there so much scalar iteration being triggered by the Oceananigans tests?

**Comment by glwagner** [comment]
There is a fair amount of scalar iteration right now, largely I think because we have array-like objects (Fields, and friends) that lack fully-featured broadcasting capabilities. This means writing something like `a .== 2` triggers scalar iteration on the GPU because it hits Base broadcasting. We _can_ fix the problem by fleshing out broadcasting a bit so `a .== 2` works "correctly" / sensibly, but we haven't prioritized it (we also didn't have broadcasting at all for Field until a month or two ago).

Possibly, if changing a global via `allowscalar(true)` were not available we would have been forced to address this deficiency in our Field abstraction sooner. So that could have been a good thing depending on your perspective. It did allow us to sweep some things under the hood. On the other hand I don't think we have any performance issues; scalar iteration is only used on very small arrays for testing where we are completely dominated by compile times.

**Pull request merged**